ResearchKit/Common/ORKLocationSelectionView.m

```diff
@@ -0,0 +1,338 @@
+    
+    NSDictionary *metrics = @{@"horizontalMargin": (_edgeToEdgeMap ? @(HorizontalTextFieldMargin) : @(0.0))};
+    [constraints addObjectsFromArray:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|-(horizontalMargin)-[_textField]-(horizontalMargin)-|" options:NSLayoutFormatDirectionLeadingToTrailing metrics:metrics views:views]];
+    
+    [NSLayoutConstraint activateConstraints:constraints];
+}
+
+- (void)setPlaceholderText:(NSString *)text {
+    _textField.placeholder = text;
+}
+
+- (void)setTextColor:(UIColor *)color {
+    _textField.textColor = color;
+}
+
+- (NSString *)enteredLocation {
+    return [_textField.text copy];
+}
+
+- (BOOL)becomeFirstResponder {
+    return [_textField becomeFirstResponder];
+}
+
+- (BOOL)isFirstResponder {
+    return [_textField isFirstResponder];
+}
+
+- (BOOL)resignFirstResponder {
+    return [_textField resignFirstResponder];
+}
+
+- (CGSize)intrinsicContentSize {
+    CGFloat height = LocationSelectionViewTextFieldHeight + (2 * LocationSelectionViewTextFieldVerticalMargin) + (_mapView.superview == nil ? 0.0 : ORKGetMetricForWindow(ORKScreenMetricLocationQuestionMapHeight, self.window));
+    return CGSizeMake(40, height);
+}
+
+- (void)showMapViewIfNecessary {
+    if (_mapView.superview) {
+        return;
+    }
+    
+    _mapView.frame = CGRectMake(0.0, 0.0, self.bounds.size.width, 0.0);
+    ORKEnableAutoLayoutForViews(@[_mapView]);
+    
+    if (!_edgeToEdgeMap) {
+        _mapView.layer.cornerRadius = 3;
+        _mapView.layer.borderColor = [UIColor colorWithRed:(204.0/255.0) green:(204.0/255.0) blue:(204.0/255.0) alpha:.75].CGColor;
+        _mapView.layer.borderWidth = 1.0 / [UIScreen mainScreen].scale;
+    }
+    [self addSubview:_mapView];
+    
+    [self loadCurrentLocationIfNecessary];
+    
+    if (_answer) {
+        [self setAnswer:_answer];
+    }
+    
+    NSMutableArray *constraints = [NSMutableArray new];
+    
+    NSDictionary *metrics = @{@"horizontalMargin": (_edgeToEdgeMap ? @(0.0) : @(HorizontalMargin))};
+    [constraints addObjectsFromArray:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|[_mapView]-(horizontalMargin)-|" options:NSLayoutFormatDirectionLeadingToTrailing metrics:metrics views:NSDictionaryOfVariableBindings(_mapView)]];
+    [constraints addObject:[NSLayoutConstraint constraintWithItem:_textField attribute:NSLayoutAttributeBottom relatedBy:NSLayoutRelationEqual toItem:_mapView attribute:NSLayoutAttributeTop multiplier:1.0 constant:-LocationSelectionViewTextFieldVerticalMargin]];
+    _mapViewHeightConstraint = [NSLayoutConstraint constraintWithItem:_mapView attribute:NSLayoutAttributeHeight relatedBy:NSLayoutRelationEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1.0 constant:ORKGetMetricForWindow(ORKScreenMetricLocationQuestionMapHeight, self.window)];
+    [constraints addObject:_mapViewHeightConstraint];
+    
+    [NSLayoutConstraint activateConstraints:constraints];
```

Comment:
`showMapViewIfNecessary` can be called again and again, more and more constraints are added?